packages/adminrouter/extra/src/test-harness/tests/test_cache.py

```diff
@@ -1005,18 +984,30 @@ def test_ip_per_task_app_with_user_networking_and_portdefinitions(
             req_data = resp.json()
             assert req_data['endpoint_id'] == 'http://127.0.0.2:80'
 
-    def test_ip_per_task_app_without_user_networking(
-            self, nginx_class, mocker, valid_user_header):
+    @pytest.mark.parametrize('host_port', ['12345', '0', None])
+    def test_app_with_container_networking_and_random_container_port(
+            self, nginx_class, mocker, valid_user_header, host_port):
+        # Testing the case when container port is specified as 0
+        # in Marathon app definition with networking mode 'container'.
+        # This means that the Marathon app container port is randomly assigned
+        # by Marathon. The chosen container port number is same as hostPort
+        # and is available through task["ports"] array. We are reusing port
+        # 16000 on 127.0.0.1 exposed by the mock server, as the one randomly
+        # chosen by Marathon.
+        # It does not matter if the host port is fixed (non-zero),
+        # randomly assigned by Marathon (0) or is not present at all:
+        # Admin Router must route the request to the specified container port.
```

Comment:
Possibly I misunderstand the meaning of this, but can you please clarify why it is necessary for AR to route to the container port? This port is internal to the container, so how can AR route to the container port?

Reply:
Every task/container  gets its own IP in `container` networking mode (exposed via `task[“ipAddresses”][1][“ipAddress”]`). On the host with that IP, the containerPort is used to reach the task. 

For an application using container networking mode (app[“networks”][1][“mode”] == “container”) with dcos overlay network (app[“networks”][1][“name”] == “dcos”) it is possible to route to the application with both <host>:<port> == task[“ipAddresses”][1][“ipAddress”] : app[“container”][“portMappings”][1][“containerPort”] as well as <host>:<port> == tasks[1][“host”] : tasks[1][“ports”][portIdx].

For a Calico-networked application, however, it is possible to route to the application only using <host>:<port> == task[“ipAddresses”][1][“ipAddress”] : app[“container”][“portMappings”][1][“containerPort”] combination. 

That combination, therefore, has been chosen for routing by AR to support both cases.

Reply:
A container port *may* be internal to the container, but AR (and Marathon) have no way of knowing that. I'm pretty sure that Marathon defaults to using *host* ip/port when in host or bridged networking mode, and container-ip:container-port when in *container* networking mode. I forget what's reported by `task["ipAddresses"][1]["ipAddress"]` - that might actually depend on the version of Marathon you're testing against, since networking changes/fixes are made from release to release.

All that said, I don't know how many of our customers use host-private container networks that aren't cluster-routeable. They don't seem very useful on a cluster like DC/OS, but I suppose for some specialized, highly secure environment there's probably an argument to be made for them. So the assumptions made here are probably fine. It might be worth documenting somewhere higher up in the docs/stack for people trying to debug service connectivity.